include/ossim/base/RWLock.h

```diff
@@ -0,0 +1,189 @@
+/**
+* This code was derived from https://gist.github.com/mshockwave
+*
+*/
+
+#ifndef ossimRWm_lockHEADER
+#define ossimRWm_lockHEADER
+
+#include <condition_variable>
+#include <mutex>
+#include <atomic>
+#include <limits>
+
+namespace ossim {
+
+    enum WaitGenre {
+        SPIN=0,
+        SLEEP=1
+    };
+    /**
+    * Code was derived from https://gist.github.com/mshockwave
+    *
+    * Has a pure c++11 implementation for read/write locks
+    * allowing one to choose the locking technique to use.
+    *
+    * You currently can choose between a SLEEP or SPIN technique
+    * for locking
+    *
+    * At the bottom we added typedefs so you do not have
+    * to specify the template values.
+    *
+    *
+    * Example:
+    * @code
+    *  ossim::RWSleepLock mutex;
+    *  // enter a section that just needs read only access 
+    *  {
+    *       ossim::ScopeReadSleepLock  lock(mutex);
+    *  }
+    *  // enter some section that requires write
+    *  {
+    *      ossim::ScopeWriteSleepLock  lock(mutex); 
+    *  }
+    * @endCode
+    */
+    template <WaitGenre WAIT_GENRE>
+    class RWLock {
+
+        std::mutex m_waitMutex;
+        std::condition_variable m_waitConditional;
+        std::atomic_int m_refCounter;
+        const int MIN_INT;
+
+    public:
+
+        RWLock();
+
+        void lockWrite();
+
+        bool tryLockWrite();
+
+        void unlockWrite();
+
+        void lockRead();
+
+        bool tryLockRead();
+
+        void unlockRead();
+    };
+
+    template <WaitGenre WAIT_GENRE>
+    RWLock<WAIT_GENRE>::RWLock() :
+            m_refCounter(0),
+            MIN_INT(std::numeric_limits<int>::min()){}
+
+    template <>
+    void RWLock<SPIN>::lockWrite() {
+
+        int expected = 0;
+        while(!m_refCounter.compare_exchange_weak(expected, MIN_INT,
+                                                  std::memory_order_acquire,
+                                                  std::memory_order_relaxed)){
+            expected = 0;
+        }
+    }
+    template <>
+    void RWLock<SLEEP>::lockWrite(){
+        int expected = 0;
+        if(!m_refCounter.compare_exchange_strong(expected, MIN_INT,
+                                                 std::memory_order_acquire,
+                                                 std::memory_order_relaxed)){
+            expected = 0;
+
+            std::unique_lock<std::mutex> lk(m_waitMutex);
+            m_waitConditional.wait(lk, [this,&expected] {
+                if(!m_refCounter.compare_exchange_strong(expected, MIN_INT,
+                                                         std::memory_order_acquire,
+                                                         std::memory_order_relaxed)){
+                    expected = 0;
+                    return false;
```

Comment:
The lockWrite() method is declared as 'void', but we have a return - is this expected?

Reply:
Hello Chris:

That is actually a lambda expression that is returning false.  The m_waitCondifitonal second argument is a lambda function.


Take care

Garrett